### ISSUE TITLE
UDP use causes libuv assertion error on Julia exit

### ISSUE THREAD

**Issue by MikeInnes** [JuliaLang/julia#13797]
``` julia
Mikes-MBP-2:~ mike$ julia -q
julia> s = UDPSocket();

julia> recv(s)
^CERROR: InterruptException:

julia> bind(s, ip"0.0.0.0", 2000)
ERROR: bind: invalid argument (EINVAL)
 in bind at socket.jl:432
 in eval at ./boot.jl:264

julia> 
Assertion failed: (!uv__io_active(&handle->io_watcher, UV__POLLIN | UV__POLLOUT)), function uv__udp_finish_close, file /Users/mike/julia/deps/srccache/libuv/src/unix/udp.c, line 65.

signal (6): Abort trap: 6
__pthread_kill at /usr/lib/system/libsystem_kernel.dylib (unknown line)
Abort trap: 6
Mikes-MBP-2:~ mike$ 
```

This reproduces on all Julia versions for me, from master way back to 0.3. Am I right in thinking that an error should be thrown on `recv` if the socket is not bound?

**Comment by ihnorton** [comment]
comment in [the code](https://github.com/JuliaLang/julia/blob/03a5b9f56d7cf33ffd3338cc7564434d87b873cd/base/socket.jl#L478-L490):
> If the socket has not been bound, it will be bound implicitly to ::0 and a random port

which matches the [upstream behavior](http://docs.libuv.org/en/v1.x/udp.html#c.uv_udp_recv_start). This is helpful for sending, but doesn't seem very useful for `recv`. We could throw an error if the socket is not open...

(if an OS-chosen free port is needed, calling `bind` on port 0 will provide that, however `getsockname` doesn't support `UDPSocket`s right now so I don't see a way to get the bound port back)

**Comment by vtjnash** [comment]
fixed by https://github.com/JuliaLang/julia/blob/master/base/socket.jl#L512

**Comment by vtjnash** [comment]
fixed by #19152 (https://github.com/JuliaLang/julia/commit/c0f824699177af6cfe2c71390400777a3febccf3#diff-1754a7e4ce910da480ee1d82ad60958fR489)

**Comment by tkelman** [comment]
Is this tested for specifically?
